File: Rust/2021/07.rs
#![feature(test)]

type Input = Vec<i32>;

fn setup(input: &str) -> Input {
    input
        .trim()
        .split(',')
        .map(|x| x.parse().unwrap())
        .collect()
}

fn part1(input: &Input) -> String {
    (0..=input.len())
        .map(|pos| input.iter().map(|n| (pos as i32 - *n).abs()).sum::<i32>())
        .min()
        .unwrap()
        .to_string()
}

fn part2(input: &Input) -> String {
    (0..=input.len())
        .map(|pos| {
            input
                .iter()
                .map(|n| {
                    let x = (pos as i32 - *n).abs();
                    (x * (x + 1)) >> 1
                })
                .sum::<i32>()
        })
        .min()
        .unwrap()
        .to_string()
}

aoc::main!(2021, 7, ex: 1);

File: Python/2021/07.py
from lib import *

input = read_input(2021, 7)


nums = ints(input)


def align1(pos):
    return sum(abs(n - pos) for n in nums)


print(min(map(align1, range(len(nums)))))


def align2(pos):
    return sum((x := abs(n - pos)) * (x + 1) // 2 for n in nums)


print(min(map(align2, range(len(nums)))))
